### PR TITLE
address errors with `-fsanitize=undefined`

### PR DESCRIPTION
**Pull request by elms**
- fix null dereferences or undefined `memcpy` calls
 - fix alignment in `myCryptoDevCb`
 - fix default dtls context assignment
 - add align configure option to force data alignment
    
TESTED:
 `./configure CFLAGS=-fsanitize=undefined --enable-all`

### CODE REVIEW

**Comment by douzzer** [comment]
more goodies lurking in there, using gcc 10.3.0 and compiler-rt-sanitizers-11.1.0-r1, with this config:

```
./configure --disable-jobserver --enable-all --enable-sp-math-all --enable-intelasm 'CFLAGS=-fsanitize=address -fsanitize=leak -fsanitize=undefined -fsanitize=float-cast-overflow -fsanitize=float-divide-by-zero -fsanitize=bounds-strict' 'LDFLAGS=-fsanitize=address -fsanitize=leak -fsanitize=undefined -fsanitize=float-cast-overflow -fsanitize=float-divide-by-zero -fsanitize=bounds-strict'
```

testsuite.test:
```
wolfssl/wolfcrypt/blake2-impl.h:89:24: runtime error: store to misaligned address 0x7ffda0da39e4 for type 'word64', which requires 8 byte alignment
0x7ffda0da39e4: note: pointer points here
  69 21 7a 30 00 00 00 00  56 08 a6 19 0f 70 b9 04  20 00 01 01 00 00 00 00  00 00 00 00 00 00 00 00
              ^ 
wolfcrypt/src/sha3.c:573:12: runtime error: load of misaligned address 0x7fff0dca0405 for type 'const word64', which requires 8 byte alignment
0x7fff0dca0405: note: pointer points here
 ae 4b f1 57 0c 41 e6  a4 0c 93 1d b2 7b 2f aa  15 a8 ce dd 52 cf f7 36  2c 4e 6e 23 da ec 0f bc  3a
             ^ 
```

unit.test:
```
   wolfSSL X509 verifsrc/wolfio.c:508:9: runtime error: member access within misaligned address 0x62c0002450c4 for type 'struct WOLFSSL_DTLS_CTX', which requires 8 byte alignment
0x62c0002450c4: note: pointer points here
  00 00 00 00 ff ff ff ff  ff ff ff ff 00 00 00 00  00 00 00 00 f4 01 00 00  00 00 00 00 00 00 09 00
              ^ 
wolfcrypt/src/sha3.c:573:12: runtime error: load of misaligned address 0x61a0002c2b59 for type 'const word64', which requires 8 byte alignment
0x61a0002c2b59: note: pointer points here
 02 02 01  01 30 05 06 03 2b 65 71  30 81 99 31 0b 30 09 06  03 55 04 06 13 02 55 53  31 10 30 0e 06
              ^ 
```

**Comment by douzzer** [comment]
oh wait, are those the 4 remaining cases?  I hadn't read your above comment before fetching your branch and kicking its tires.

is there something about those 4 that makes them hard to fix?  it would be lovely to be able to add `-fsanitize=undefined`.

**Comment by elms** [comment]
@douzzer Those are the remaining. I couldn't get to the bottom of them in the time I allocated and didn't want to put off the other fixes. I think they are `malloc`ed blocks and somehow not being aligned. Maybe another set of eyes, or I can revisit next week.

**Comment by douzzer** [comment]
I'll take a few minutes on them and see what I can see.  Agreed let's not let these 4 hold up the substantial improvements already in the PR.

**Comment by douzzer** [comment]
Candidate fixes for these last 4 (3 unique) are in commit 15484558 at
https://github.com/douzzer/wolfssl/tree/fsanitize/undefined_fixes

The gnarliest was the one in api.c.  It wasn't just a benign-ish unaligned access error.  Turns out a pointer wasn't being properly initialized, in a way that could lead to undetected buffer overruns and data corruption.

File: wolfcrypt/src/hc128.c
Comment (elms) [inline wolfcrypt/src/hc128.c:280]: `memcpy` wrap in `XSTREAM_ALIGN`?

File: wolfcrypt/src/rabbit.c
Comment (elms) [inline wolfcrypt/src/rabbit.c:154]: `memcpy` wrap in `XSTREAM_ALIGN`?

File: wolfcrypt/src/sha3.c
Comment (elms) [inline wolfcrypt/src/sha3.c:575]: Should this be conditional on `XSTREAM_ALIGN`?

File: wolfssl/wolfcrypt/blake2-impl.h
Comment (elms) [inline wolfssl/wolfcrypt/blake2-impl.h:89]: `memcpy` wrap in `XSTREAM_ALIGN`?

Comment (SparkiDev) [reply wolfssl/wolfcrypt/blake2-impl.h:89]: Replace with aligned assignments like in sha3.c:Load64BitBigEndian().
Obviously this is a little endian assignment so is a little different.

File: wolfcrypt/src/hc128.c
Comment (SparkiDev) [reply wolfcrypt/src/hc128.c:280]: Alignment should have been fixed in wc_Hc128_SetKey().

File: wolfcrypt/src/rabbit.c
Comment (SparkiDev) [reply wolfcrypt/src/rabbit.c:154]: Alignment was fixed in wc_RabbitSetKey().

File: wolfcrypt/src/sha3.c
Comment (SparkiDev) [reply wolfcrypt/src/sha3.c:575]: XSTREAM_ALIGN informs WOLFSSL_GENERAL_ALIGNMENT.

File: wolfssl/wolfcrypt/blake2-impl.h
Comment (SparkiDev) [reply wolfssl/wolfcrypt/blake2-impl.h:89]: Don't use XMEMCPY.
Check if alignment is large enough for this path or whether the byte-by-byte is to be used.
Could also have multiple alignment paths like SHA-3 code.

File: wolfcrypt/src/sha3.c
Comment (SparkiDev) [reply wolfcrypt/src/sha3.c:575]: WOLFSSL_GENERAL_ALIGNMENT may be 0 which I take to mean: "no alignment issues".

File: wolfssl/wolfcrypt/blake2-impl.h
Comment (elms) [reply wolfssl/wolfcrypt/blake2-impl.h:89]: To have just a single implementation for those operations would it make sense to live in `misc.c`?

File: wolfcrypt/src/sha3.c
Comment (elms) [reply wolfcrypt/src/sha3.c:575]: Ahh I see. With `XSTREAM_ALIGN` I get ed448 sign and verify reports. Should the alignment of data be handled be the caller or in `ed448.c`?

<details>
<summary>ed448_verify_msg</summary>

```
wolfcrypt/src/sha3.c:559:19: runtime error: load of misaligned address 0x5605ddd4e9b9 for type 'word3
2', which requires 4 byte alignment
0x5605ddd4e9b9: note: pointer points here
 02 02 01  01 30 05 06 03 2b 65 71  30 81 99 31 0b 30 09 06  03 55 04 06 13 02 55 53  31 10 30 0e 06
              ^ 
    #0 0x7f6d75398ecc in Load64BitBigEndian wolfcrypt/src/sha3.c:559
    #1 0x7f6d75398ecc in Sha3Update wolfcrypt/src/sha3.c:635
    #2 0x7f6d755642c4 in ed448_verify_msg wolfcrypt/src/ed448.c:413
    #3 0x7f6d755654f8 in wc_ed448_verify_msg wolfcrypt/src/ed448.c:467
    #4 0x7f6d753e7f30 in ConfirmSignature wolfcrypt/src/asn.c:7926
    #5 0x7f6d753ea1fd in ParseCertRelative wolfcrypt/src/asn.c:10389
    #6 0x7f6d755dd03c in ProcessPeerCertParse src/internal.c:11094
    #7 0x7f6d7565a886 in ProcessPeerCerts src/internal.c:11514
    #8 0x7f6d7566152f in DoCertificate src/internal.c:12500
    #9 0x7f6d7566152f in DoHandShakeMsgType src/internal.c:13415
    #10 0x7f6d756707e4 in DoHandShakeMsg src/internal.c:13627
    #11 0x7f6d756707e4 in ProcessReply src/internal.c:16333
    #12 0x7f6d756b92ee in wolfSSL_connect src/ssl.c:12744
    #13 0x5605dc13620a in client_test examples/client/client.c:3232
    #14 0x5605dc123db4 in execute_test_case tests/suites.c:551
    #15 0x5605dc1250cf in test_harness tests/suites.c:727
    #16 0x5605dc125d63 in SuiteTest tests/suites.c:905
    #17 0x5605dc057161 in unit_test tests/unit.c:85
    #18 0x5605dc057161 in unit_test tests/unit.c:46
    #19 0x7f6d7445a0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)

```

</details>

<details>
<summary>ed448_sign_msg</summary>

```
wolfcrypt/src/sha3.c:560:19: runtime error: load of misaligned address 0x7ffd5183f609 for type 'word32', which requires 4 byte alignment
0x7ffd5183f609: note: pointer points here
 0c 41 e6  a4 0c 93 1d b2 7b 2f aa  15 a8 ce dd 52 cf f7 36  2c 4e 6e 23 da ec 0f bc  3a 79 b6 80 6e
              ^ 
    #0 0x7f318cf97fd5 in Load64BitBigEndian wolfcrypt/src/sha3.c:560
    #1 0x7f318cf97fd5 in Sha3Update wolfcrypt/src/sha3.c:635
    #2 0x7f318d162e19 in ed448_sign_msg wolfcrypt/src/ed448.c:207
    #3 0x7f318d164417 in wc_ed448_sign_msg wolfcrypt/src/ed448.c:280
    #4 0x5571e81f7931 in ed448_test wolfcrypt/test/test.c:26838
    #5 0x5571e81ff055 in wolfcrypt_test wolfcrypt/test/test.c:1275
    #6 0x5571e821947b in testsuite_test testsuite/testsuite.c:153
    #7 0x7f318c0590b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #8 0x5571e81ae02d in _start (/home/elms/work/wolf/wolfssl/testsuite/.libs/testsuite.test+0x3b02d)

```

</details>

File: wolfssl/wolfcrypt/blake2-impl.h
Comment (SparkiDev) [reply wolfssl/wolfcrypt/blake2-impl.h:89]: Yes. We have code for big-endian operations in misc.c already (not optimized).

**Comment by elms** [comment]
@SparkiDev I've made some changes and rebased this PR. Two things I have mixed feelings about and would appreciate your input on:
 1. New configure option that defaults alignment to yes
 2. Still needing to set alignment to `1` to pass without error. I'd like it to pass by default.

File: wolfcrypt/src/sha3.c
Comment (SparkiDev) [reply wolfcrypt/src/sha3.c]: What compiler are you using?
I can't get reports with gcc9 or clang10 on my Linux box.

Comment (elms) [reply wolfcrypt/src/sha3.c]: I used gcc 9.3.0 on a linux box.

```
./configure CFLAGS=-fsanitize=undefined\ -g --enable-all
UBSAN_OPTIONS=print_stacktrace=1:log_path=./UBSAN make test
```

This generates the two logs `UBSAN.<pid>` where pid was the pid of the process for testsuite and unit.test. I used `-g` to get the lines in the tracebacks.

Comment (SparkiDev) [reply wolfcrypt/src/sha3.c]: Able to reproduce now.
Change the loading of the data (sha3.c:599) to call Load64Unaligned():
```
WC_STATIC WC_INLINE word64 Load64Unaligned(unsigned char *a)
{
    return ((word64)a[0] <<  0) | 
           ((word64)a[1] <<  8) | 
           ((word64)a[2] << 16) | 
           ((word64)a[3] << 24) | 
           ((word64)a[4] << 32) | 
           ((word64)a[5] << 40) | 
           ((word64)a[6] << 48) | 
           ((word64)a[7] << 56);
}
```
gcc converts this to a single instruction.

Comment (elms) [reply wolfcrypt/src/sha3.c]: Thanks, the line number didn't match but think I changed it as you were suggesting.

Performance on my machine seem to have negligible impact. Are there specific configs that may impact this more?

<details>
<summary>head of tree</summary>

```
$ ./config.status --c
'--enable-all'
$ git describe 
v4.7.1r-599-gae2f2b246
$ ./wolfcrypt/benchmark/benchmark -csv -rabbit -hc128 -sha -sha2 -sha224 -sha256 -sha384 -sha512 -sha3
------------------------------------------------------------------------------
 wolfSSL version 4.7.1
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
This format allows you to easily copy the output to a csv file.

Symmetric Ciphers:

Algorithm,MB/s,Cycles per byte,
HC128,1712.890,1.89,
RABBIT,700.209,4.63,
SHA,513.502,6.31,
SHA-224,217.334,14.92,
SHA-256,220.317,14.71,
SHA-384,336.544,9.63,
SHA-512,338.646,9.57,
SHA3-224,307.666,10.54,
SHA3-256,291.198,11.13,
SHA3-384,220.581,14.70,
SHA3-512,154.931,20.92,
Benchmark complete

```

</details>

<details>
<summary>This PR</summary>

```
$ ./config.status --c
'--enable-all'
$ git describe 
v4.7.1r-601-gbcee1dbf6
$ ./wolfcrypt/benchmark/benchmark -csv -rabbit -hc128 -sha -sha2 -sha224 -sha256 -sha384 -sha512 -sha3
------------------------------------------------------------------------------
 wolfSSL version 4.7.1
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
This format allows you to easily copy the output to a csv file.

Symmetric Ciphers:

Algorithm,MB/s,Cycles per byte,
HC128,1681.548,1.93,
RABBIT,712.736,4.55,
SHA,515.687,6.29,
SHA-224,221.840,14.61,
SHA-256,221.584,14.63,
SHA-384,337.009,9.62,
SHA-512,339.915,9.54,
SHA3-224,309.002,10.49,
SHA3-256,291.040,11.14,
SHA3-384,224.874,14.41,
SHA3-512,154.027,21.05,
Benchmark complete
```


</details>

Comment (SparkiDev) [reply wolfcrypt/src/sha3.c]: No other configs. Just wanted you to check against master.
Looks fine.

**Comment by SparkiDev** [comment]
Disable aligndata by default.
Use the following macro in rabbit.c and hc128.c:
#define LOAD_LE32(a)           \
    (((word32)(a)[0] <<  0) |  \
     ((word32)(a)[1] <<  8) |  \
     ((word32)(a)[2] << 16) |  \
     ((word32)(a)[3] << 24))

**Comment by elms** [comment]
> Don't define XSTREAM_ALIGN in aligndata.
> Use the following macro in rabbit.c and hc128.c:
> #define LOAD_LE32(a)
> (((word32)(a)[0] << 0) |
> ((word32)(a)[1] << 8) |
> ((word32)(a)[2] << 16) |
> ((word32)(a)[3] << 24))

@SparkiDev I'm not sure I understand your suggestion. I understand you want me to change `configure.ac` to remove setting `XSTREAM_ALIGN` and add a new macro but it's not clear what you mean beyond that. To avoid additional back and forth can I ask you just make the change and push directly to the PR?

**Review by douzzer — CHANGES_REQUESTED** [review]
So close!  Reviewed all the changes again line by line, all looks good.

With gcc-10.3.0 -fsanitize=undefined, the only errors left are
```
In file included from /usr/include/string.h:519,
                 from ./wolfssl/wolfcrypt/types.h:519,
                 from ./wolfssl/internal.h:28,
                 from src/internal.c:73:
In function ‘strncpy’,
    inlined from ‘SetCipherList’ at src/internal.c:20481:9:
/usr/include/bits/string_fortified.h:95:10: error: ‘__builtin_strncpy’ specified bound 49 equals destination size [-Werror=stringop-truncation]
   95 |   return __builtin___strncpy_chk (__dest, __src, __len,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   96 |       __glibc_objsize (__dest));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/bits/string_fortified.h:95:10: error: ‘__builtin_strncpy’ specified bound 49 equals destination size [-Werror=stringop-truncation]
In file included from /usr/include/string.h:519,
                 from ./wolfssl/wolfcrypt/types.h:519,
                 from ./wolfssl/internal.h:28,
                 from src/ssl.c:37:
In function ‘strncpy’,
    inlined from ‘wolfSSL_X509_VERIFY_PARAM_set1_ip_asc’ at src/ssl.c:26315:13,
    inlined from ‘wolfSSL_X509_VERIFY_PARAM_set1_ip_asc’ at src/ssl.c:26305:5:
/usr/include/bits/string_fortified.h:95:10: error: ‘__builtin_strncpy’ specified bound 46 equals destination size [-Werror=stringop-truncation]
   95 |   return __builtin___strncpy_chk (__dest, __src, __len,
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   96 |       __glibc_objsize (__dest));
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~
```

Feel free to push my fix for these:
```
diff --git a/src/internal.c b/src/internal.c
index 03a61f728..2d9877e7e 100644
--- a/src/internal.c
+++ b/src/internal.c
@@ -20470,7 +20470,7 @@ int SetCipherList(WOLFSSL_CTX* ctx, Suites* suites, const char* list)
         word32 length;
 
         next = XSTRSTR(next, ":");
-        length = MAX_SUITE_NAME + 1;
+        length = MAX_SUITE_NAME;
         if (next != NULL) {
             word32 currLen = (word32)(next - current);
             if (length > currLen) {
diff --git a/src/ssl.c b/src/ssl.c
index 98de82ae8..f7e46a66d 100644
--- a/src/ssl.c
+++ b/src/ssl.c
@@ -26312,7 +26312,7 @@ int wolfSSL_X509_VERIFY_PARAM_set1_ip_asc(WOLFSSL_X509_VERIFY_PARAM *param,
             param->ipasc[0] = '\0';
         }
         else {
-            XSTRNCPY(param->ipasc, ipasc, WOLFSSL_MAX_IPSTR);
+            XSTRNCPY(param->ipasc, ipasc, WOLFSSL_MAX_IPSTR-1);
             param->ipasc[WOLFSSL_MAX_IPSTR-1] = '\0';
         }
         ret = WOLFSSL_SUCCESS;
```

and then send it back over to me.

File: src/ssl.c
Comment (elms) [inline src/ssl.c:26570]: @SparkiDev Was this intentional? Looks like gcc-10 flags it and @douzzer is suggesting adding back the `-1`.

**Comment by elms** [comment]
@douzzer Thanks for the thorough review. Changes are rebased and compiles and runs with gcc 10 
now with no undefined reports.

We need to get gcc-10 added to PRB tests. The change in `internal.c` was already present in head of the project.

**Review by douzzer — APPROVED** [review]
Nice last-minute catch on the `LITTLE32` -> `LOAD_LE32` refactor in `DoKey()`!

**Pull request merged**
